File: aite/gateway/schema.py
"""按 `ToolSpec.parameters` 校验 arguments（owner: T3）。

§3.2 的调用顺序里第三步是「按 ToolSpec.parameters 校验 arguments」。`parameters`
是 JSON Schema，但 §3.1 冻结的那 5 个 `GATEWAY_TOOLS` 只用到了它极小的一个子集：
`type` / `properties` / `required` / `default` / `minimum` / `maximum` /
`minItems` / `maxItems` / `items` / `enum`。

所以这里手写这个子集，不引 jsonschema —— §3.0 的依赖是冻结的，加一条要停下来报告，
而为了校验五个两三个字段的 schema 去动冻结文件不值得。支持的关键字之外的一律**忽略**
（不静默放过整个字段，只是不额外约束），这样以后 spec 往 schema 里加关键字时，
最坏情况是校验偏松，不会变成「认不出就整个拒掉」。

除 spec 写的之外多了一条本层策略：**不认识的参数直接拒**。JSON Schema 的默认语义是
允许额外字段，但这是给模型用的工具入口，模型把参数名写错时早点收到 invalid_args
比静默丢掉那个参数好。
"""
from typing import Any

__all__ = ["SchemaViolation", "validate_arguments"]


class SchemaViolation(ValueError):
    """arguments 不合 schema。Gateway 把它翻成 ToolErrorCode.invalid_args。"""


_TYPE_PREDICATES: dict[str, Any] = {
    # bool 是 int 的子类，必须显式排掉，否则 True 会被当成合法的 integer。
    "boolean": lambda v: isinstance(v, bool),
    "integer": lambda v: isinstance(v, int) and not isinstance(v, bool),
    "number": lambda v: isinstance(v, int | float) and not isinstance(v, bool),
    "string": lambda v: isinstance(v, str),
    "array": lambda v: isinstance(v, list),
    "object": lambda v: isinstance(v, dict),
    "null": lambda v: v is None,
}


def validate_arguments(schema: dict[str, Any], arguments: dict[str, Any]) -> dict[str, Any]:
    """校验并补默认值，返回可以直接喂给工具实现的 arguments。

    只做校验和填 `default`，不做类型转换 —— 字符串 "50" 不会被悄悄变成整数 50，
    模型给错类型就该拿到 invalid_args。
    """
    if not isinstance(arguments, dict):
        raise SchemaViolation(f"arguments 必须是对象，收到 {type(arguments).__name__}")

    declared_type = schema.get("type", "object")
    if declared_type != "object":
        raise SchemaViolation(f"工具的 parameters 顶层必须是 object，schema 写的是 {declared_type!r}")

    properties: dict[str, Any] = schema.get("properties") or {}
    required: list[str] = list(schema.get("required") or [])

    unknown = sorted(set(arguments) - set(properties))
    if unknown:
        known = sorted(properties) or ["（这个工具不接受任何参数）"]
        raise SchemaViolation(f"不认识的参数 {unknown}；可用的是 {known}")

    missing = [name for name in required if name not in arguments]
    if missing:
        raise SchemaViolation(f"缺少必填参数 {missing}")

    out: dict[str, Any] = {}
    for name, subschema in properties.items():
        if name in arguments:
            out[name] = _check(arguments[name], subschema, name)
        elif "default" in subschema:
            out[name] = subschema["default"]
    return out


def _check(value: Any, subschema: dict[str, Any], path: str) -> Any:
    if not isinstance(subschema, dict):        # schema 本身坏了，不是调用方的错
        raise SchemaViolation(f"{path} 的 schema 不是对象：{subschema!r}")

    expected = subschema.get("type")
    types = [expected] if isinstance(expected, str) else list(expected or [])
    if types:
        if any(t not in _TYPE_PREDICATES for t in types):
            raise SchemaViolation(f"{path} 的 schema 用了不支持的 type：{expected!r}")
        if not any(_TYPE_PREDICATES[t](value) for t in types):
            raise SchemaViolation(f"{path} 应该是 {'/'.join(types)}，收到 {_name_of(value)}")

    if "enum" in subschema and value not in subschema["enum"]:
        raise SchemaViolation(f"{path} 只能是 {subschema['enum']} 之一，收到 {value!r}")

    if isinstance(value, str):
        _check_string(value, subschema, path)
    if isinstance(value, int | float) and not isinstance(value, bool):
        _check_number(value, subschema, path)
    if isinstance(value, list):
        return _check_array(value, subschema, path)
    if isinstance(value, dict):
        return _check_object(value, subschema, path)
    return value


def _check_string(value: str, subschema: dict[str, Any], path: str) -> None:
    low = subschema.get("minLength")
    high = subschema.get("maxLength")
    if low is not None and len(value) < low:
        raise SchemaViolation(f"{path} 至少要 {low} 个字符，收到 {len(value)} 个")
    if high is not None and len(value) > high:
        raise SchemaViolation(f"{path} 最多 {high} 个字符，收到 {len(value)} 个")


def _check_number(value: float, subschema: dict[str, Any], path: str) -> None:
    checks = (
        ("minimum", lambda lo: value < lo, "不能小于"),
        ("maximum", lambda hi: value > hi, "不能大于"),
        ("exclusiveMinimum", lambda lo: value <= lo, "必须大于"),
        ("exclusiveMaximum", lambda hi: value >= hi, "必须小于"),
    )
    for key, violated, phrase in checks:
        bound = subschema.get(key)
        if bound is not None and violated(bound):
            raise SchemaViolation(f"{path} {phrase} {bound}，收到 {value}")


def _check_array(value: list[Any], subschema: dict[str, Any], path: str) -> list[Any]:
    low = subschema.get("minItems")
    high = subschema.get("maxItems")
    if low is not None and len(value) < low:
        raise SchemaViolation(f"{path} 至少要 {low} 项，收到 {len(value)} 项")
    if high is not None and len(value) > high:
        raise SchemaViolation(f"{path} 最多 {high} 项，收到 {len(value)} 项")
    items = subschema.get("items")
    if not isinstance(items, dict):
        return value
    return [_check(item, items, f"{path}[{i}]") for i, item in enumerate(value)]


def _check_object(value: dict[str, Any], subschema: dict[str, Any], path: str) -> dict[str, Any]:
    properties = subschema.get("properties")
    if not isinstance(properties, dict):
        return value
    missing = [n for n in (subschema.get("required") or []) if n not in value]
    if missing:
        raise SchemaViolation(f"{path} 缺少必填字段 {missing}")
    out = dict(value)
    for name, sub in properties.items():
        if name in value:
            out[name] = _check(value[name], sub, f"{path}.{name}")
        elif "default" in sub:
            out[name] = sub["default"]
    return out


def _name_of(value: Any) -> str:
    return "null" if value is None else f"{type(value).__name__}({value!r})"[:80]

File: core/crates/gateway/src/schema.rs
//! 按 `ToolSpec.parameters` 校验 arguments（对应旧 `aite/gateway/schema.py`）。
//!
//! §3.2 的调用顺序里第三步是「按 ToolSpec.parameters 校验 arguments」。校验本体交给
//! `jsonschema` crate（workspace 已钉 0.56），这一层只做三件旧实现里的本层策略：
//!
//! 1. **未知参数直接拒**（JSON Schema 默认允许额外字段；这是给模型用的工具入口，
//!    参数名写错时早点收到 invalid_args 比静默丢掉好）。消息里列出可用参数名，模型能照着改。
//! 2. **补 `default`**，返回可以直接喂给工具实现的 arguments（嵌套对象里的 default 也补）。
//! 3. **不做类型转换**，而且比 JSON Schema 更严一条：`integer` 只认整数字面量。
//!    draft 2020-12 把 `50.0` 当合法 integer，旧实现的 `isinstance(v, int)` 不认 ——
//!    这里补一道严格检查，保持与 Python 版逐条一致（`bool` 不是 integer 由 JSON 类型系统天然保证）。
//!
//! 错误消息沿用旧实现的中文措辞：它会原样进 `ToolResult.content` 给模型看。
use jsonschema::error::{TypeKind, ValidationErrorKind};
use jsonschema::paths::LocationSegment;
use serde_json::{Map, Value};

/// arguments 不合 schema。Gateway 把它翻成 `ToolErrorCode::InvalidArgs`。
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct SchemaViolation(pub String);

impl SchemaViolation {
    fn of(message: impl Into<String>) -> Self {
        Self(message.into())
    }
}

/// 校验并补默认值，返回可以直接喂给工具实现的 arguments。
///
/// 顺序与旧实现一致：顶层 schema 形状 → 未知参数 → 缺必填 → 逐字段校验 → 补 default。
/// 顺序要紧：打错一个参数名时，「不认识的参数」比「缺少必填参数」更能让模型改对。
pub fn validate_arguments(
    schema: &Value,
    arguments: &Map<String, Value>,
) -> Result<Map<String, Value>, SchemaViolation> {
    match schema.get("type") {
        None => {}
        Some(Value::String(t)) if t == "object" => {}
        Some(other) => {
            return Err(SchemaViolation::of(format!(
                "工具的 parameters 顶层必须是 object，schema 写的是 {other}"
            )));
        }
    }

    let empty = Map::new();
    let properties = schema
        .get("properties")
        .and_then(Value::as_object)
        .unwrap_or(&empty);

    let mut unknown: Vec<&str> = arguments
        .keys()
        .filter(|k| !properties.contains_key(k.as_str()))
        .map(String::as_str)
        .collect();
    if !unknown.is_empty() {
        unknown.sort_unstable();
        let mut known: Vec<&str> = properties.keys().map(String::as_str).collect();
        known.sort_unstable();
        let known = if known.is_empty() {
            "[\"（这个工具不接受任何参数）\"]".to_string()
        } else {
            render_list(&known)
        };
        return Err(SchemaViolation::of(format!(
            "不认识的参数 {}；可用的是 {known}",
            render_list(&unknown)
        )));
    }

    let required: Vec<&str> = schema
        .get("required")
        .and_then(Value::as_array)
        .map(|items| items.iter().filter_map(Value::as_str).collect())
        .unwrap_or_default();
    let missing: Vec<&str> = required
        .iter()
        .copied()
        .filter(|name| !arguments.contains_key(*name))
        .collect();
    if !missing.is_empty() {
        return Err(SchemaViolation::of(format!(
            "缺少必填参数 {}",
            render_list(&missing)
        )));
    }

    let validator = jsonschema::validator_for(schema)
        .map_err(|e| SchemaViolation::of(format!("工具的 parameters schema 有问题：{e}")))?;
    let instance = Value::Object(arguments.clone());
    if let Some(error) = validator.iter_errors(&instance).next() {
        return Err(SchemaViolation::of(render_violation(&error)));
    }

    match walk(&instance, schema, "")? {
        Value::Object(map) => Ok(map),
        // instance 是上面亲手包的 Object，walk 不改变种类。
        other => Err(SchemaViolation::of(format!(
            "校验后的 arguments 不是对象：{other}"
        ))),
    }
}

/// 递归补 `default` + 严格 integer 检查。返回新值，不原地改。
fn walk(value: &Value, schema: &Value, path: &str) -> Result<Value, SchemaViolation> {
    check_strict_integer(value, schema, path)?;
    match value {
        Value::Object(map) => {
            let Some(properties) = schema.get("properties").and_then(Value::as_object) else {
                return Ok(value.clone());
            };
            let mut out = map.clone();
            for (name, subschema) in properties {
                let child = if path.is_empty() {
                    name.clone()
                } else {
                    format!("{path}.{name}")
                };
                if let Some(given) = map.get(name) {
                    out.insert(name.clone(), walk(given, subschema, &child)?);
                } else if let Some(default) = subschema.get("default") {
                    out.insert(name.clone(), default.clone());
                }
            }
            Ok(Value::Object(out))
        }
        Value::Array(items) => {
            let Some(subschema) = schema.get("items").filter(|s| s.is_object()) else {
                return Ok(value.clone());
            };
            let mut out = Vec::with_capacity(items.len());
            for (i, item) in items.iter().enumerate() {
                out.push(walk(item, subschema, &format!("{path}[{i}]"))?);
            }
            Ok(Value::Array(out))
        }
        _ => Ok(value.clone()),
    }
}

/// `integer` 只认整数字面量：`50.0` 在 JSON Schema 里算 integer，旧实现不算。
fn check_strict_integer(value: &Value, schema: &Value, path: &str) -> Result<(), SchemaViolation> {
    let types = declared_types(schema);
    if !types.contains(&"integer") || types.contains(&"number") {
        return Ok(());
    }
    if let Value::Number(n) = value
        && n.as_i64().is_none()
        && n.as_u64().is_none()
    {
        return Err(SchemaViolation::of(format!(
            "{} 应该是 {}，收到 {}",
            display_path(path),
            types.join("/"),
            received(value)
        )));
    }
    Ok(())
}

fn declared_types(schema: &Value) -> Vec<&str> {
    match schema.get("type") {
        Some(Value::String(t)) => vec![t.as_str()],
        Some(Value::Array(items)) => items.iter().filter_map(Value::as_str).collect(),
        _ => Vec::new(),
    }
}

fn render_violation(error: &jsonschema::ValidationError<'_>) -> String {
    let raw_path = python_path(error.instance_path());
    let path = display_path(&raw_path);
    let instance = error.instance();
    match error.kind() {
        ValidationErrorKind::Type {
            kind: TypeKind::Single(expected),
        } => format!("{path} 应该是 {expected}，收到 {}", received(instance)),
        ValidationErrorKind::Type {
            kind: TypeKind::Multiple(expected),
        } => {
            let names: Vec<String> = expected.into_iter().map(|t| t.to_string()).collect();
            format!(
                "{path} 应该是 {}，收到 {}",
                names.join("/"),
                received(instance)
            )
        }
        ValidationErrorKind::Enum { options } => {
            format!("{path} 只能是 {options} 之一，收到 {instance}")
        }
        ValidationErrorKind::Minimum { limit } => {
            format!("{path} 不能小于 {limit}，收到 {instance}")
        }
        ValidationErrorKind::Maximum { limit } => {
            format!("{path} 不能大于 {limit}，收到 {instance}")
        }
        ValidationErrorKind::ExclusiveMinimum { limit } => {
            format!("{path} 必须大于 {limit}，收到 {instance}")
        }
        ValidationErrorKind::ExclusiveMaximum { limit } => {
            format!("{path} 必须小于 {limit}，收到 {instance}")
        }
        ValidationErrorKind::MinLength { limit } => format!(
            "{path} 至少要 {limit} 个字符，收到 {} 个",
            char_count(instance)
        ),
        ValidationErrorKind::MaxLength { limit } => format!(
            "{path} 最多 {limit} 个字符，收到 {} 个",
            char_count(instance)
        ),
        ValidationErrorKind::MinItems { limit } => {
            format!("{path} 至少要 {limit} 项，收到 {} 项", item_count(instance))
        }
        ValidationErrorKind::MaxItems { limit } => {
            format!("{path} 最多 {limit} 项，收到 {} 项", item_count(instance))
        }
        ValidationErrorKind::Required { property } => {
            format!("{path} 缺少必填字段 [{property}]")
        }
        _ => format!("{path} 不合 schema：{error}"),
    }
}

/// jsonschema 的 instance_path 是 JSON Pointer（`/who/id`、`/items/0`）；
/// 旧实现的消息用 `who.id` / `items[0]`，照它渲染，模型看到的字样才一致。
fn python_path(location: &jsonschema::paths::Location) -> String {
    let mut out = String::new();
    for segment in location.segments() {
        match segment {
            LocationSegment::Property(name) => {
                if out.is_empty() {
                    out.push_str(&name);
                } else {
                    out.push('.');
                    out.push_str(&name);
                }
            }
            LocationSegment::Index(i) => {
                out.push_str(&format!("[{i}]"));
            }
        }
    }
    out
}

fn display_path(path: &str) -> &str {
    if path.is_empty() { "arguments" } else { path }
}

/// 与旧实现的 `_name_of` 同一用途：把收到的值连类型一起说出来（截到 80 字符）。
fn received(value: &Value) -> String {
    if value.is_null() {
        return "null".to_string();
    }
    let kind = match value {
        Value::Bool(_) => "boolean",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
        Value::Null => "null",
    };
    let text = format!("{kind}({value})");
    text.chars().take(80).collect()
}

fn char_count(value: &Value) -> usize {
    value.as_str().map(|s| s.chars().count()).unwrap_or(0)
}

fn item_count(value: &Value) -> usize {
    value.as_array().map(Vec::len).unwrap_or(0)
}

fn render_list(items: &[&str]) -> String {
    let quoted: Vec<String> = items.iter().map(|s| format!("\"{s}\"")).collect();
    format!("[{}]", quoted.join(", "))
}
